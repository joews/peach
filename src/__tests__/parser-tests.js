/* eslint-env jest */
const parse = require('../parser')
const { fixture } = require('./helpers')

//
// snapshot tests for the parser
//

function testFixture (fixtureName) {
  test(fixtureName, () => {
    expect(parse(fixture(fixtureName))).toMatchSnapshot()
  })
}

function testParse (peachCode) {
  test(peachCode, () => {
    expect(parse(peachCode)).toMatchSnapshot()
  })
}

testFixture('fibonacci.peach')
testFixture('str.peach')
testFixture('tail-recursion.peach')

testParse('(true => 1)')
testParse('([true|tail] => 1)')
