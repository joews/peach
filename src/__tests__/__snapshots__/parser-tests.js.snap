exports[`test ((true|tail) => 1) 1`] = `
Array [
  Object {
    "clauses": Array [
      Object {
        "body": Object {
          "type": "Numeral",
          "value": 1,
        },
        "pattern": Array [
          Object {
            "head": Object {
              "type": "Bool",
              "value": true,
            },
            "tail": Object {
              "name": "tail",
              "type": "Name",
            },
            "type": "DestructuredList",
          },
        ],
      },
    ],
    "type": "Fn",
  },
]
`;

exports[`test (true => 1) 1`] = `
Array [
  Object {
    "clauses": Array [
      Object {
        "body": Object {
          "type": "Numeral",
          "value": 1,
        },
        "pattern": Array [
          Object {
            "type": "Bool",
            "value": true,
          },
        ],
      },
    ],
    "type": "Fn",
  },
]
`;

exports[`test fibonacci.peach 1`] = `
Array [
  Object {
    "name": "fib",
    "type": "Def",
    "value": Object {
      "clauses": Array [
        Object {
          "body": Object {
            "type": "Numeral",
            "value": 1,
          },
          "pattern": Array [
            Object {
              "type": "Numeral",
              "value": 0,
            },
          ],
        },
        Object {
          "body": Object {
            "type": "Numeral",
            "value": 1,
          },
          "pattern": Array [
            Object {
              "type": "Numeral",
              "value": 1,
            },
          ],
        },
        Object {
          "body": Object {
            "type": "List",
            "values": Array [
              Object {
                "name": "+",
                "type": "Name",
              },
              Object {
                "type": "List",
                "values": Array [
                  Object {
                    "name": "fib",
                    "type": "Name",
                  },
                  Object {
                    "type": "List",
                    "values": Array [
                      Object {
                        "name": "-",
                        "type": "Name",
                      },
                      Object {
                        "name": "x",
                        "type": "Name",
                      },
                      Object {
                        "type": "Numeral",
                        "value": 1,
                      },
                    ],
                  },
                ],
              },
              Object {
                "type": "List",
                "values": Array [
                  Object {
                    "name": "fib",
                    "type": "Name",
                  },
                  Object {
                    "type": "List",
                    "values": Array [
                      Object {
                        "name": "-",
                        "type": "Name",
                      },
                      Object {
                        "name": "x",
                        "type": "Name",
                      },
                      Object {
                        "type": "Numeral",
                        "value": 2,
                      },
                    ],
                  },
                ],
              },
            ],
          },
          "pattern": Array [
            Object {
              "name": "x",
              "type": "Name",
            },
          ],
        },
      ],
      "type": "Fn",
    },
  },
  Object {
    "type": "List",
    "values": Array [
      Object {
        "name": "map",
        "type": "Name",
      },
      Object {
        "name": "fib",
        "type": "Name",
      },
      Object {
        "isQuoted": true,
        "type": "List",
        "values": Array [
          Object {
            "type": "Numeral",
            "value": 0,
          },
          Object {
            "type": "Numeral",
            "value": 1,
          },
          Object {
            "type": "Numeral",
            "value": 2,
          },
          Object {
            "type": "Numeral",
            "value": 3,
          },
          Object {
            "type": "Numeral",
            "value": 4,
          },
          Object {
            "type": "Numeral",
            "value": 5,
          },
          Object {
            "type": "Numeral",
            "value": 6,
          },
          Object {
            "type": "Numeral",
            "value": 7,
          },
          Object {
            "type": "Numeral",
            "value": 8,
          },
          Object {
            "type": "Numeral",
            "value": 9,
          },
          Object {
            "type": "Numeral",
            "value": 10,
          },
        ],
      },
    ],
  },
]
`;

exports[`test str.peach 1`] = `
Array [
  Object {
    "isQuoted": true,
    "type": "List",
    "values": Array [
      Object {
        "type": "Str",
        "value": "str.peach before escape\\\`and after!",
      },
      Object {
        "type": "Str",
        "value": "hey ðŸŒŸ",
      },
      Object {
        "type": "Str",
        "value": "multi
      line
  string",
      },
      Object {
        "type": "Str",
        "value": "such\\\\\\\\\\\`escape",
      },
      Object {
        "type": "Str",
        "value": "with	tab",
      },
      Object {
        "type": "Str",
        "value": "new
line",
      },
      Object {
        "type": "Str",
        "value": "back\\slash",
      },
      Object {
        "type": "Str",
        "value": "back\\\\to\\back",
      },
    ],
  },
]
`;

exports[`test tail-recursion.peach 1`] = `
Array [
  Object {
    "name": "factorial",
    "type": "Def",
    "value": Object {
      "clauses": Array [
        Object {
          "body": Object {
            "name": "n",
            "type": "Name",
          },
          "pattern": Array [
            Object {
              "name": "n",
              "type": "Name",
            },
            Object {
              "type": "Numeral",
              "value": 1,
            },
          ],
        },
        Object {
          "body": Object {
            "type": "List",
            "values": Array [
              Object {
                "name": "factorial",
                "type": "Name",
              },
              Object {
                "type": "List",
                "values": Array [
                  Object {
                    "name": "*",
                    "type": "Name",
                  },
                  Object {
                    "name": "n",
                    "type": "Name",
                  },
                  Object {
                    "name": "x",
                    "type": "Name",
                  },
                ],
              },
              Object {
                "type": "List",
                "values": Array [
                  Object {
                    "name": "-",
                    "type": "Name",
                  },
                  Object {
                    "name": "x",
                    "type": "Name",
                  },
                  Object {
                    "type": "Numeral",
                    "value": 1,
                  },
                ],
              },
            ],
          },
          "pattern": Array [
            Object {
              "name": "n",
              "type": "Name",
            },
            Object {
              "name": "x",
              "type": "Name",
            },
          ],
        },
      ],
      "type": "Fn",
    },
  },
  Object {
    "name": "f",
    "type": "Def",
    "value": Object {
      "clauses": Array [
        Object {
          "body": Object {
            "type": "List",
            "values": Array [
              Object {
                "name": "factorial",
                "type": "Name",
              },
              Object {
                "type": "Numeral",
                "value": 1,
              },
              Object {
                "name": "n",
                "type": "Name",
              },
            ],
          },
          "pattern": Array [
            Object {
              "name": "n",
              "type": "Name",
            },
          ],
        },
      ],
      "type": "Fn",
    },
  },
  Object {
    "type": "List",
    "values": Array [
      Object {
        "name": "f",
        "type": "Name",
      },
      Object {
        "type": "Numeral",
        "value": 32768,
      },
    ],
  },
]
`;
